webstruct/html_tokenizer.py

```diff
@@ -167,13 +172,46 @@ def tokenize_single(self, tree):
 
         """
         tree = copy.deepcopy(tree)
-        self.sequence_encoder.reset()
         self._prepare_tree(tree)
-        res = list(zip(*self._process_tree(tree)))
-        if not res:
+        tokens_tree = list(self._process_tree(tree))
+        if not tokens_tree:
             return [], []
+        res = self._encode(tokens_tree)
+
         return list(res[0]), list(res[1])
 
+    def _encode(self, tokens_tree):
+        chains, tags = [], []
+        self.sequence_encoder.reset()
+        for node_tokens, tree, is_tail in tokens_tree:
```

Comment:
I think calling self.sequence_encoder.encode, etc. multiple times in a loop is a wrong thing to do. All information is already available, so there is only a little benefit in processing data in chunks. A better way to do it is to combine all sequences, and then pass them to a sequence encoder once. It would simplify sequence encoder API a lot, and allow to implement BilouEncoder.